### PR TITLE
ENH: Add da.tril and da.triu

### PR DESCRIPTION
**Pull request by sinhrks**
Added ``da.triu`` and ``da.tril``. As a first impl, input must be a square matrix with equal horizontal / vertical chunks (can have remainder).

```
import numpy as np
import dask.array as da

A = np.random.random_integers(0, 10, (6, 6))
dA = da.from_array(A, (2, 2))

da.triu(dA).compute()
# array([[9, 1, 4, 5, 9, 6],
#        [0, 6, 2, 6, 5, 7],
#        [0, 0, 5, 3, 5, 5],
#        [0, 0, 0, 4, 3, 8],
#        [0, 0, 0, 0, 9, 3],
#        [0, 0, 0, 0, 0, 6]])

da.triu(dA, 3).compute()
# array([[ 0.,  0.,  0.,  5.,  9.,  6.],
#        [ 0.,  0.,  0.,  0.,  5.,  7.],
#        [ 0.,  0.,  0.,  0.,  0.,  5.],
#        [ 0.,  0.,  0.,  0.,  0.,  0.],
#        [ 0.,  0.,  0.,  0.,  0.,  0.],
#        [ 0.,  0.,  0.,  0.,  0.,  0.]])
```

### CODE REVIEW

File: dask/array/tests/test_array_core.py [hunk 1]
```diff
@@ -1781,3 +1781,25 @@ def test_to_imperative():
...
+    A = np.random.randn(20, 20)
+    for chunk in [5, 7]:
+        dA = da.from_array(A, (chunk, chunk))
+
+        assert np.allclose(da.triu(dA).compute(), np.triu(A))
+        assert np.allclose(da.tril(dA).compute(), np.tril(A))
+
+        for k in range(-25, 25):
+            assert np.allclose(da.triu(dA, k).compute(), np.triu(A, k))
+            assert np.allclose(da.tril(dA, k).compute(), np.tril(A, k))
```
Comment (mrocklin) [inline dask/array/tests/test_array_core.py]: Are all of these 100 cases necessary to ensure correctness?  Is there a smaller set that would do as well?

It is useful to keep test times fast if possible.

File: dask/array/core.py [hunk 2]
```diff
@@ -2660,6 +2660,106 @@ def block_hist(x, weights=None):
...
+    --------
+    triu : upper triangle of an array
+    """
+    if m.ndim != 2:
+        raise ValueError('input must be 2 dimensional')
+    if m.shape[0] != m.shape[1]:
+        raise NotImplementedError('input must be a square matrix')
+    if m.chunks[0][0] != m.chunks[1][0]:
+        msg = ('chunks must be a square. '
+               'Use .rechunk method to change the size of chunks.')
```
Comment (mrocklin) [inline dask/array/core.py]: This test and this message don't match exactly.  I think that the following is a counter-example.

    ((2, 3, 4), (2, 4, 3))

Your test with `set` from the LU PR seemed accurate to me.

File: dask/array/core.py [hunk 3]
```diff
@@ -2660,6 +2660,106 @@ def block_hist(x, weights=None):
...
+    for i in range(rdim):
+        for j in range(hdim):
+            if chunk * (j - i + 1) < k:
+                dsk[(name, i, j)] = (m.name, i, j)
+            elif chunk * (j - i - 1) < k <= chunk * (j - i + 1):
+                dsk[(name, i, j)] = (np.tril, (m.name, i, j), k - (chunk * (j - i)))
+            else:
+                dsk[(name, i, j)] = (np.zeros, (m.chunks[0][i], m.chunks[1][j]))
+    dsk.update(m.dask)
+    return Array(dsk, name, shape=m.shape, chunks=(m.chunks[0], m.chunks[1]))
```
Comment (mrocklin) [inline dask/array/core.py]: can this be `chunks=m.chunks`?

**Comment by mrocklin** [comment]
This looks good.  I had only a couple of very minor comments.

File: dask/array/core.py (hunk 2 above)
Comment (sinhrks) [inline dask/array/core.py]: Thanks, correct. Is there anywhere I can put this type of checks as utility funcs?

Comment (mrocklin) [inline dask/array/core.py]: Leaving it as is seems best to me.  It's not a common check within dask.array.

**Comment by sinhrks** [comment]
Thanks, rebased.

**Pull request merged**